### ISSUE TITLE
Downstream user requested QuickStart [Debian bug #927085]

### ISSUE THREAD

**Issue by sten0** [jorgenschaefer/elpy#1596]
Hi,

A Debian user filed a bug requesting a QuickStart doc, either as a section or standalone (at any rate, probably docs/quickstart.rst).  User appears to be coming from a Python, but not Emacs background.  It might also be nice to provide a corresponding docs/refcard.rst for users of similar background.  Maybe a refcard.rst with the most used shortcuts/functions `highlighted` would do the trick?

1) User requests a short list of the most used functions and their shortcuts.
   * User requests highlighting of `elpy-shell-send-region-or-buffer` in particular
   * docs/ide.rst already provides this info in comprehensive form (P.S. that's what I prefer, thank you!)
2) User requests an `add-hook` emacs.el hint for `elpy-shell-switch-to-shell`.

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=927085

Thanks!
Nicholas

**Comment by galaunay** [comment]
Hi, 
Thanks for forwarding this.

I added a `quickstart` section in the doc in PR #1598 (any feedback on that is welcome).
The documentation on readthedoc is stuck to elpy 1.27.0 for now, so it may take a bit of time to be updated.

About the hook, I guess you are refering to this:
> Do you happen to know if it is possible to configure emacs to open a
Python buffer in elpy mode by default and skip the C-c C-z?

My answer would be that you don't need to open the shell with `C-c C-z` before evaluating the code with `C-c C-c`. `C-c C-c` will automatically open a python shell if needed. you can still open a shell automatically when opening a python file with `(add-hook 'elpy-mode-hook 'elpy-shell-display-buffer t)`

> 6) Your bug report template should include the output of
> elpy-config. I've included mine at the bottom. I'm not sure what is going on with
>
>    Elpy..............: Not found (Python), 1.28.0 (Emacs Lisp)
>
>though. What's the difference between those two cases?

This had been fixed in Elpy 1.29.0.

> > 3) I suggest mentioning the upstream manual in README.Debian. Namely,
> > https://elpy.readthedocs.io/en/latest/
> >
> 
> Unfortunately this will not provide accurate documentation for users
> of Debian buster in the coming months, because that URL will document
> a newer version of Elpy, and buster is shipping with 1.28.0.  Also, a
> link is useless for users who are offline.

I think we could provide documentation for all Elpy versions on readthedoc, so you could link to the version of the doc debian is actually shipping.
Let me know if it would be useful.

**Comment by sten0** [comment]
Hi,

On Thu, May 09, 2019 at 12:54:16PM -0700, galaunay wrote:
> Hi, 
> Thanks for forwarding this.
>

Thanks for the quick reply! :-)  I'm hoping that the Debian release
team will grant permission to make some fixups for its Elpy package
that will ship in Buster (Debian 10) for the best out-of-the-box
Elpy experience.

> I added a `quickstart` section in the doc in PR #1598 (any feedback on that is welcome).
> The documentation on readthedoc is stuck to elpy 1.27.0 for now, so it may take a bit of time to be updated.
>

Thanks, I'll comment there momentarily, I'll cherry pick the commits
once they're merged to master.

> About the hook, I guess you are refering to this:
> > Do you happen to know if it is possible to configure emacs to open a
> Python buffer in elpy mode by default and skip the C-c C-z?
> 
> My answer would be that you don't need to open the shell with `C-c C-z` before evaluating the code with `C-c C-c`. `C-c C-c` will automatically open a python shell if needed. you can still open a shell automatically when opening a python file with `(add-hook 'elpy-mode-hook 'elpy-shell-display-buffer t)`
>

I agree, and prefer the default behaviour :-)

For #1598: I think it would be valuable to mention this in the
QuickStart.  Providing a hint for the hook allows a user with Python
background who is new to Emacs to immediately get a sense of
familiarity when that user is coming from Eclipse+PyDev, Spyder,
PyCharm, Thonny, et al, which have a persistent shell frame.  Also, it
seems like a good idea to mention this hook isn't necessary, and also
maybe link to a doc that explains how Emacs conventions are
different/better than the static layout used by most IDEs.

> > 6) Your bug report template should include the output of
> > elpy-config. I've included mine at the bottom. I'm not sure what is going on with
> >
> >    Elpy..............: Not found (Python), 1.28.0 (Emacs Lisp)
> >
> >though. What's the difference between those two cases?
> 
> This had been fixed in Elpy 1.29.0.
>

Sorry, this one was my fault.  I didn't document with sufficient
clarity the fact that the Debian package doesn't build the Python 2
bits.  tldr, it would have been a lot of work to enable Python 2
support in the Debian CI instance for Elpy, and there was talk of
cutting Python 2 from Buster, so I focused on making sure everything
was 100% good with Python 3 (August 2018).

Sadly we're stuck with 1.28.0 for Buster/Debian 10 (uploads of new
versions are prohibited), but if there are any high priority commits
you think should be cherry picked, please let me know.

Out-of-the-box Debian's package is configured to use /usr/bin/python3,
and that line is:
```
Elpy..............: 1.28.0
```

This user had customised `elpy-rpc-pythonpath` to `python`.

> > > 3) I suggest mentioning the upstream manual in README.Debian. Namely,
> > > https://elpy.readthedocs.io/en/latest/
> > >
> > 
> > Unfortunately this will not provide accurate documentation for users
> > of Debian buster in the coming months, because that URL will document
> > a newer version of Elpy, and buster is shipping with 1.28.0.  Also, a
> > link is useless for users who are offline.
> 
> I think we could provide documentation for all Elpy versions on readthedoc, so you could link to the version of the doc debian is actually shipping.
> Let me know if it would be useful.
>

After consulting some people on the topic of html docs in Debian, it
became clear the package should probably include them, because a lot
users look for html docs first, rather than man/info/Emacs built-in
docs; I'll generated and bundle these docs in the next upload.  The
(Debian perspective) ideal is to not depend on the net for
documentation.

Handling this on the Debian side means readthedocs can stay simple and
track the latest tagged release.

Oh, but readthedocs is currently used as the homepage of the Debian
package...I can change that to the github project if you prefer.  I
guess the question can be reduced to: Do you want users to treat
readthedocs or github as the homepage?  If it's readthedocs, then yes,
all Elpy versions should be documented, but if the github project is
suitable as a user-facing homepage then there's no need.

BTW, I'm excited because the user-feedback in that Debian bug might be
indicative of the following: dev with Python background -> Elpy ->
Emacs -> use Emacs for other languages in education/career :-)

Cheers!
Nicholas

**Comment by galaunay** [comment]
> For #1598: I think it would be valuable to mention this in the
QuickStart.  Providing a hint for the hook allows a user with Python
background who is new to Emacs to immediately get a sense of
familiarity when that user is coming from Eclipse+PyDev, Spyder,
PyCharm, Thonny, et al, which have a persistent shell frame.  Also, it
seems like a good idea to mention this hook isn't necessary, and also
maybe link to a doc that explains how Emacs conventions are
different/better than the static layout used by most IDEs.

I understand your point, but it seems very specific to people starting to use Emacs. 
I think the quickstart should remain more generic.
What do you think about adding a paragraph on this in the relevant part of the doc (at the start of the `Shell buffer` or `Interactive python` sections) ?

> Sadly we're stuck with 1.28.0 for Buster/Debian 10 (uploads of new
versions are prohibited), but if there are any high priority commits
you think should be cherry picked, please let me know.
>
> Out-of-the-box Debian's package is configured to use /usr/bin/python3,
and that line is:
>```
>Elpy..............: 1.28.0
>```

This is just a display issue in `elpy-config`, this is fortunately not affecting anything else.
So not a big deal I guess.

> Oh, but readthedocs is currently used as the homepage of the Debian
package...I can change that to the github project if you prefer.  I
guess the question can be reduced to: Do you want users to treat
readthedocs or github as the homepage?  If it's readthedocs, then yes,
all Elpy versions should be documented, but if the github project is
suitable as a user-facing homepage then there's no need.

I guess it makes sense to have readthedoc as homepage.
I am pretty sure it is doable to generate the documentation for every version of Elpy.
The problem is that Jorgen is the only one to have admin rights on readthedoc at the moment and he has been away for a while.
I'll fix that as soon as he re-appears.

> BTW, I'm excited because the user-feedback in that Debian bug might be
indicative of the following: dev with Python background -> Elpy ->
Emacs -> use Emacs for other languages in education/career :-)

That's nice, I am very happy that Elpy is appealing enough to make people give Emacs a try.
I am pretty sure they wont regret it :).

**Comment by sten0** [comment]
@galaunay thanks again for your work! :-)

> I understand your point, but it seems very specific to people starting to use Emacs.
> I think the quickstart should remain more generic.
> What do you think about adding a paragraph on this in the relevant part of the doc (at the start of the Shell buffer or Interactive python sections) ? 

That's a good idea.  It seems wise in the sense that the quickstart should be like an appetiser--enough to whet the appetite, but not a tldr the fine manual thing  ;-)

One minor issue I encountered while testing #1598 with unconfigured Emacs (to simulate a new user's experience), following only the README and this section, is that following the docs is not always sufficient to activate Elpy in Python buffers.  Steps to reproduce:

1. Open a python file
2. M-x elpy-enable
3. (Elpy is not enabled in the buffer) C-RET is undefined
4. `M-x elpy-mode` needs to be explicitly enabled.

I'm not sure if a note to the affect of "elpy-enable should be run before opening a python file, otherwise it may be necessary to run `M-x elpy-mode`" should go in README.md or here.  Yeah, arguably it's hand-holding, but bug reports like "elpy-enable didn't enable Elpy in the python buffer I already had open, even though python-mode works fine" are no fun, and couple of words to note this seems easy.

Off to sleep,
Nicholas

**Comment by galaunay** [comment]
> I'm not sure if a note to the affect of "elpy-enable should be run before opening a python file, otherwise it may be necessary to run M-x elpy-mode" should go in README.md or here. Yeah, arguably it's hand-holding, but bug reports like "elpy-enable didn't enable Elpy in the python buffer I already had open, even though python-mode works fine" are no fun, and couple of words to note this seems easy.

You are right, it may be a bit confusing.
I modified `elpy-enable` to make sure elpy is activated in every python buffer (PR #1602).
Shouldn't be a problem anymore :).


I will try to think about what we can do documentation-wise for new Emacs users.
Honestly, the best would be a screencast or youtube tutorial, but that's a bit outside of my skills.

**Comment by sten0** [comment]
> You are right, it may be a bit confusing.
I modified elpy-enable to make sure elpy is activated in every python buffer (PR #1602).
Shouldn't be a problem anymore :).

Oh, good call!  I agree, that's a much better way to resolve this.  I'm still waiting for an ACK on the unblock request, but will cherry pick this commit and ping the release team.

I'll open a new issue to discuss the tutorial.

**Comment by galaunay** [comment]
Hi sten0,

I regained control on readthedoc.
The documentation for Elpy 1.28.0 is now available [here](https://elpy.readthedocs.io/en/1.28.0/).
(It is available for other versions as well if you need them).

**Comment by sten0** [comment]
Hi @galaunay,

Thanks!  Looks good :-)  Update on the unblock:  it was successful, so users of Debian stable have the QuickStart.  I think that's everything for this bug, so I'm closing it.  Please see my two comments on your documentation overhaul PR.
